admin-dev/themes/new-theme/js/pages/product/components/combination-list/bulk-edition-handler.ts

```diff
@@ -64,46 +74,52 @@ export default class BulkEditionHandler {
   }
 
   private init(): void {
+    this.getBulkEditFormBtn()!.addEventListener('click', () => this.showFormModal());
+  }
+
+  private getBulkEditFormBtn(): HTMLButtonElement | null {
     const bulkEditionFormBtn = this.tabContainer.querySelector<HTMLButtonElement>(CombinationMap.bulkCombinationFormBtn);
 
     if (!(bulkEditionFormBtn instanceof HTMLButtonElement)) {
       console.error(`${CombinationMap.bulkCombinationFormBtn} was expected to be HTMLButtonElement`);
-      return;
+      return null;
     }
+
+    return bulkEditionFormBtn;
+  }
+
+  private async showFormModal(): Promise<void> {
+    const bulkEditionFormBtn = this.getBulkEditFormBtn()!;
     const {modalConfirmLabel, modalCancelLabel} = bulkEditionFormBtn.dataset;
     const {formUrl} = bulkEditionFormBtn.dataset;
 
     if (typeof formUrl !== 'string') {
       console.error('Mandatory attribute "data-form-url" is missing');
       return;
     }
-
-    bulkEditionFormBtn.addEventListener('click', () => this.showFormModal(
-      formUrl,
-      bulkEditionFormBtn.innerHTML,
-      modalConfirmLabel || 'Confirm',
-      modalCancelLabel || 'Cancel',
-    ));
-  }
-
-  private async showFormModal(
-    formUrl: string,
-    modalTitle: string,
-    confirmButtonLabel: string,
-    closeButtonLabel: string,
-  ): Promise<void> {
     const selectedCombinationIds = await this.bulkChoicesSelector.getSelectedIds();
     const selectedCombinationsCount = selectedCombinationIds.length;
     let initialSerializedData: string;
-    const iframeModal = new FormIframeModal({
+
+    const iframeModal: FormIframeModal = new FormIframeModal({
       id: CombinationMap.bulkFormModalId,
-      modalTitle,
+      modalTitle: bulkEditionFormBtn.innerHTML,
       formUrl,
       autoSizeContainer: 'form[name="bulk_combination"]',
       closable: true,
-      confirmButtonLabel: confirmButtonLabel.replace(/%combinations_number%/, String(selectedCombinationsCount)),
-      closeButtonLabel,
+      confirmButtonLabel: modalConfirmLabel?.replace(/%combinations_number%/, String(selectedCombinationsCount)),
+      closeButtonLabel: modalCancelLabel,
       onFormLoaded: (form: HTMLFormElement) => {
+        if (this.violatedFormContent) {
+          // Replace current form content with new one fetched from api which has violations rendered
+          // eslint-disable-next-line no-param-reassign
+          form.parentElement!.innerHTML = this.violatedFormContent.trim();
+
+          // Select the freshly replaced form in DOM or else event listeners wouldn't work
+          // eslint-disable-next-line no-param-reassign
+          form = <HTMLFormElement>iframeModal.modal.iframe.contentDocument!.querySelector(CombinationMap.bulkForm);
```

Comment:
Please use the generic instead of casting, it breaks the whole typing, then handle the fact that the form could be undefined even if it loaded (because you can receive an error in the iframe)

Reply:
```suggestion
          form = iframeModal.modal.iframe.contentDocument.querySelector<HTMLFormElement>(CombinationMap.bulkForm);
```

That's what the generic looks like, and indeed since the HTML has been replaced it's safer to check if the replaced content indeed contains the targeted form